core/miner/main/main.cpp

```diff
@@ -328,14 +340,94 @@ namespace fc {
     auto window{mining::WindowPoStScheduler::create(
         napi, manager, manager, *config.actor)};
 
+    auto graphsync{std::make_shared<storage::ipfs::graphsync::GraphsyncImpl>(
+        host, scheduler)};
+    graphsync->start();
+    auto datatransfer{data_transfer::DataTransfer::make(host, graphsync)};
+
+    auto markets_ipld{std::make_shared<storage::ipfs::LeveldbDatastore>(
+        prefixed("markets_ipld/"))};
+    auto gs_sub{graphsync->subscribe([&](auto &, auto &data) {
+      OUTCOME_EXCEPT(markets_ipld->set(data.cid, data.content));
+    })};
+    auto stored_ask{std::make_shared<markets::storage::provider::StoredAsk>(
+        prefixed("stored_ask/"), napi, *config.actor)};
+    auto piece_storage{std::make_shared<storage::piece::PieceStorageImpl>(
+        prefixed("storage_provider/"))};
+    auto sector_blocks{std::make_shared<sectorblocks::SectorBlocksImpl>(miner)};
+    auto chain_events{
+        std::make_shared<markets::storage::chain_events::ChainEventsImpl>(
+            napi)};
+    OUTCOME_TRY(chain_events->init());
+    auto piece_io{std::make_shared<markets::pieceio::PieceIOImpl>(
+        markets_ipld, config.join("piece_io"))};
+    auto filestore{std::make_shared<storage::filestore::FileSystemFileStore>()};
+    auto storage_provider{
+        std::make_shared<markets::storage::provider::StorageProviderImpl>(
+            minfo.seal_proof_type,
+            host,
+            markets_ipld,
+            datatransfer,
+            stored_ask,
+            io,
+            piece_storage,
+            napi,
+            sector_blocks,
+            chain_events,
+            *config.actor,
+            piece_io,
+            filestore)};
+    OUTCOME_TRY(storage_provider->init());
+    auto retrieval_provider{
+        std::make_shared<markets::retrieval::provider::RetrievalProviderImpl>(
+            host,
+            datatransfer,
+            napi,
+            piece_storage,
+            markets_ipld,
+            one_key("retrieval_provider_ask", leveldb),
+            manager,
+            miner)};
+    retrieval_provider->start();
+
     auto mapi{std::make_shared<api::Api>()};
+    mapi->DealsImportData = [&](auto &proposal, auto &path) {
```

Comment:
Maybe we should move api initiation in another file, as it is with node api

Reply:
It's miner api